### DISCUSSION TITLE
Array-like type support

### DISCUSSION THREAD

**Discussion by kevinsantana11** [Cloud-Drift/clouddrift#507] [Ideas]
[Original Thread](https://github.com/Cloud-Drift/clouddrift/pull/504#issue-2446770974)
> Hi @kevinsantana11 / @selipot, I tried to fix `mypy` for all modules and failed (to say the least). I did this PR as a first step, we have a few arguments that were:
> 
> ```
> x: int | None = 0
> ```
> 
> which should simply be:
> 
> ```
> x: int = 0
> ```
> 
> To make typing a bit more easy for us, I think we should discuss a strategy. As of right now, we have a lot of functions that try to accept all types:
> 
> ```
> list | np.ndarray | pd.Series | xr.DataArray
> ```
> 
> but this is not constant across `clouddrift`, so for example, another function will have:
> 
> ```
> np.ndarray[float] | pd.Series | xr.DataArray
> ```
> 
> so if we pass a list somewhere `mypy` will complain.
> 
> Similarly, we are accepting `list` at a few places alongside other `array_like` structures and do operations like (if `x` is `array_like`):
> 
> ```
> x / CONSTANT
> x.shape
> x.size
> ```
> 
> which are not valid if `isinstance(x, list)`.
> 
> My main questions are:
> 
>     * do we really want to support all the types?
> 
>     * what do other libraries do in that situation? Or someone knows what's the best practices here?
> 
> 
> One solution could be to defined a `clouddrift_array_like_type` once, and then reused it.. Something we do in `pairs.py` with:
> 
> ```
> array_like = Union[list, np.ndarray, pd.Series, xr.DataArray]
> ```
> 
> Maybe we can copy this to a discussion?

---

## Main Question
* do we really want to support all the types?
* what do other libraries do in that situation? Or someone knows what's the best practices here?

**Comment by kevinsantana11** [comment]
I need to dig into how other libraries handle this but I think its a good idea to standardize on the types we support. 

We should also consider supporting DaskArrays since it can help enable parallelization throughout the library but am not sure on supporting vanilla python lists.

**Reply by selipot** [reply]
I rarely, if ever, use python lists to represent vectors of data so i think it would be ok for me not to include those in the ArrayLike type.
